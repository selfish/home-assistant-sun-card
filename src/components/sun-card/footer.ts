import { html, TemplateResult } from "lit";

import { ESunCardI18NKeys, ISunCardConfig, TSunCardData, TSunCardFields, TSunCardTimes } from "../../types";
import { Helpers } from "../../utils/helpers";
import { I18N } from "../../utils/i18n";

export class Footer {
  private data: TSunCardData;
  private i18n: I18N;
  private times: TSunCardTimes;
  private fields: TSunCardFields;

  constructor(config: ISunCardConfig, data: TSunCardData) {
    this.data = data;

    this.i18n = config.i18n!;
    this.times = data?.times;
    this.fields = config.fields!;
  }

  public render(): TemplateResult {
    return html`
      <div class="sun-card-footer">
        <div class="sun-card-field-row">
          ${
      this.fields?.dawn !== undefined && this.times?.dawn !== undefined
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Dawn, this.times.dawn)
        : Helpers.nothing()
    }
          ${
      this.fields?.noon !== undefined && this.times?.noon !== undefined
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Noon, this.times.noon)
        : Helpers.nothing()
    }
          ${
      this.fields?.dusk !== undefined && this.times?.dusk !== undefined
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Dusk, this.times.dusk)
        : Helpers.nothing()
    }
        </div>

        <div class="sun-card-field-row">
          ${
      this.fields?.azimuth !== undefined && this.data?.azimuth !== undefined
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Azimuth, this.data?.azimuth)
        : Helpers.nothing()
    }
          ${
      this.fields?.elevation !== undefined && this.data?.elevation !== undefined
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Elevation, this.data?.elevation)
        : Helpers.nothing()
    }
        </div>
      </div>
    `;
  }
}
