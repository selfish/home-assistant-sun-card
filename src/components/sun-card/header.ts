import { html, TemplateResult } from "lit";

import { ESunCardI18NKeys, ISunCardConfig, TSunCardData, TSunCardFields, TSunCardTimes } from "../../types";
import { Helpers } from "../../utils/helpers";
import { I18N } from "../../utils/i18n";

export class Header {
  private title?: string;
  private times: TSunCardTimes;
  private fields: TSunCardFields;
  private i18n: I18N;

  constructor(config: ISunCardConfig, data: TSunCardData) {
    this.title = config.title;
    this.fields = config.fields!;
    this.times = data?.times;

    this.i18n = config.i18n!;
  }

  public render(): TemplateResult {
    return html`
      ${this.showTitle() ? this.renderTitle() : Helpers.nothing()}
      ${this.renderHeader()}
    `;
  }

  private renderTitle(): TemplateResult {
    return html`<div class="sun-card-title">${this.title}</div>`;
  }

  private renderHeader(): TemplateResult {
    return html`
      <div class="sun-card-header">
        ${
      this.fields?.sunrise && this.times?.sunrise
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Sunrise, this.times.sunrise)
        : Helpers.nothing()
    }
        ${
      this.fields?.sunset && this.times?.sunset
        ? Helpers.renderFieldElement(this.i18n, ESunCardI18NKeys.Sunset, this.times.sunset)
        : Helpers.nothing()
    }
      </div>
    `;
  }

  private showTitle(): boolean {
    return this.title !== undefined;
  }
}
