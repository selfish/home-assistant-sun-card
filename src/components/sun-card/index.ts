export * from "./card-content";
export * from "./footer";
export * from "./header";
export * from "./sun-card";
export * from "./sun-curve";
