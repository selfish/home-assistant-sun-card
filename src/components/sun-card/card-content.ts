import { html, TemplateResult } from "lit";

import { ISunCardConfig, TSunCardData } from "../../types";
import { Helpers } from "../../utils/helpers";
import { Footer } from "./footer";
import { Header } from "./header";
import { SunCurve } from "./sun-curve";

export class CardContent {
  private config: ISunCardConfig;
  private data: TSunCardData;

  constructor(config: ISunCardConfig, data: TSunCardData) {
    this.config = config;
    this.data = data;
  }

  render(): TemplateResult {
    return html`
      <ha-card>
        <div class="sun-card ${this.config.darkMode ? "sun-card-dark" : ""}">
          ${this.showHeader() ? this.renderHeader() : Helpers.nothing()}
          ${this.renderGraph()}
          ${this.showFooter() ? this.renderFooter() : Helpers.nothing()}
        </div>
      </ha-card>
    `;
  }

  private renderHeader(): TemplateResult {
    return new Header(this.config, this.data).render();
  }

  private renderGraph(): TemplateResult {
    return new SunCurve(this.data).render();
  }

  private renderFooter(): TemplateResult {
    return new Footer(this.config, this.data).render();
  }

  private showHeader(): boolean {
    // logic based on config
    return true;
  }

  private showFooter(): boolean {
    // logic based on config
    return true;
  }
}
