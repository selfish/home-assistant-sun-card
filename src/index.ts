import "./components/sun-card";
import "./components/editor";
