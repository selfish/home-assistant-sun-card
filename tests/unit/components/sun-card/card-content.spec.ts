import { html } from "lit";

import { CardContent } from "../../../../src/components/sun-card/card-content";
import { ISunCardConfig, TSunCardData } from "../../../../src/types";
import { Helpers } from "../../../../src/utils/helpers";
import { CustomSnapshotSerializer, TemplateResultTestHelper } from "../../../helpers";

jest.mock("../../../../src/components/sun-card/header", () => require("../../../mocks/header"));
jest.mock("../../../../src/components/sun-card/sun-curve", () => require("../../../mocks/sun-curve"));
jest.mock("../../../../src/components/sun-card/footer", () => require("../../../mocks/footer"));

expect.addSnapshotSerializer(new CustomSnapshotSerializer());

describe("CardContent", () => {
  describe("render", () => {
    beforeAll(() => {
      jest.spyOn(Helpers, "nothing").mockImplementation(() => html``);
    });

    it("sets dark mode when it is set to true on the config", async () => {
      const config: ISunCardConfig = {
        type: "sun-card",
        darkMode: true
      };

      const sunCardContent = new CardContent(config, {} as TSunCardData);
      const element = window.document.createElement("test-element") as TemplateResultTestHelper<typeof sunCardContent.render>;
      element.templateResultFunction = () => sunCardContent.render();
      window.document.body.appendChild(element);
      await element.updateComplete;

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
    });

    it("does not set dark mode when it is set to false on the config", async () => {
      const config: ISunCardConfig = {
        type: "sun-card",
        darkMode: false
      };

      const sunCardContent = new CardContent(config, {} as TSunCardData);
      const element = window.document.createElement("test-element") as TemplateResultTestHelper<typeof sunCardContent.render>;
      element.templateResultFunction = () => sunCardContent.render();
      window.document.body.appendChild(element);
      await element.updateComplete;

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
    });

    it("renders the header when show header returns true", async () => {
      const sunCardContent = new CardContent({} as ISunCardConfig, {} as TSunCardData);
      jest.spyOn((sunCardContent as unknown as { showHeader: () => boolean }), "showHeader").mockImplementation(() => true);

      const element = window.document.createElement("test-element") as TemplateResultTestHelper<typeof sunCardContent.render>;
      element.templateResultFunction = () => sunCardContent.render();
      window.document.body.appendChild(element);
      await element.updateComplete;

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
    });

    it("does not render the header when show header returns false", async () => {
      const sunCardContent = new CardContent({} as ISunCardConfig, {} as TSunCardData);
      jest.spyOn((sunCardContent as unknown as { showHeader: () => boolean }), "showHeader").mockImplementation(() => false);

      const element = window.document.createElement("test-element") as TemplateResultTestHelper<typeof sunCardContent.render>;
      element.templateResultFunction = () => sunCardContent.render();
      window.document.body.appendChild(element);
      await element.updateComplete;

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
    });

    it("renders the footer when show footer returns true", async () => {
      const sunCardContent = new CardContent({} as ISunCardConfig, {} as TSunCardData);
      jest.spyOn((sunCardContent as unknown as { showFooter: () => boolean }), "showFooter").mockImplementation(() => true);

      const element = window.document.createElement("test-element") as TemplateResultTestHelper<typeof sunCardContent.render>;
      element.templateResultFunction = () => sunCardContent.render();
      window.document.body.appendChild(element);
      await element.updateComplete;

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
    });

    it("does not render the footer when show footer returns false", async () => {
      const sunCardContent = new CardContent({} as ISunCardConfig, {} as TSunCardData);
      jest.spyOn((sunCardContent as unknown as { showFooter: () => boolean }), "showFooter").mockImplementation(() => false);

      const element = window.document.createElement("test-element") as TemplateResultTestHelper<typeof sunCardContent.render>;
      element.templateResultFunction = () => sunCardContent.render();
      window.document.body.appendChild(element);
      await element.updateComplete;

      expect(element.shadowRoot!.innerHTML).toMatchSnapshot();
    });
  });
});
