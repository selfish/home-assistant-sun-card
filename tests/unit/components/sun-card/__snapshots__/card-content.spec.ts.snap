// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardContent render does not render the footer when show footer returns false 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit--><!--?-->
        </div>
      </ha-card>
    
`;

exports[`CardContent render does not render the header when show header returns false 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit--><!--?-->
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`CardContent render does not set dark mode when it is set to false on the config 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`CardContent render renders the footer when show footer returns true 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`CardContent render renders the header when show header returns true 1`] = `
<!---->
      <ha-card>
        <div class="sun-card ">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;

exports[`CardContent render sets dark mode when it is set to true on the config 1`] = `
<!---->
      <ha-card>
        <div class="sun-card sun-card-dark">
          <!--?lit-->
      <div class="sun-card-header">
        SUN CARD HEADER
      </div>
    
          <!--?lit-->
      <div class="sun-card-graph">
        SUN CARD GRAPH
      </div>
    
          <!--?lit-->
      <div class="sun-card-footer">
        SUN CARD FOOTER
      </div>
    
        </div>
      </ha-card>
    
`;
